src/runtime/BackingStore.h

```diff
@@ -22,53 +22,166 @@
 
 class BackingStore : public gc {
-    friend class ArrayBufferObject;
+public:
+    static BackingStore* createDefaultNonSharedBackingStore(size_t byteLength);
+    static BackingStore* createNonSharedBackingStore(void* data, size_t byteLength, BackingStoreDeleterCallback callback, void* callbackData);
+
+#if defined(ENABLE_THREADING)
+    static BackingStore* createDefaultSharedBackingStore(size_t byteLength);
+    // SharedBackingStore is newly created by sharing with already created one
+    static BackingStore* createSharedBackingStore(SharedDataBlockInfo* sharedInfo);
+#endif
+
+    // indicates whether the backing store was created for SharedArrayBuffer (SharedDataBlock)
+    virtual bool isShared() const = 0;
+    virtual void* data() const = 0;
+    virtual size_t byteLength() const = 0;
+    virtual void* deleterData() const = 0;
+    virtual SharedDataBlockInfo* sharedDataBlockInfo() const = 0;
+    virtual void reallocate(size_t newByteLength) = 0;
+
+    void* operator new(size_t size) = delete;
+    void* operator new[](size_t size) = delete;
+};
+
+class NonSharedBackingStore : public BackingStore {
+    friend class BackingStore;
 
 public:
-    BackingStore(size_t byteLength);
-    BackingStore(void* data, size_t byteLength, BackingStoreDeleterCallback callback,
-                 void* callbackData, bool isShared = false);
-    ~BackingStore();
+    virtual bool isShared() const override
+    {
+        return false;
+    }
 
-    void* data() const
+    virtual void* data() const override
     {
         return m_data;
     }
 
-    size_t byteLength() const
+    virtual size_t byteLength() const override
     {
         return m_byteLength;
     }
 
-    bool isShared() const
+    virtual void* deleterData() const override
     {
-        return m_isShared;
+        return m_deleterData;
     }
 
-    void* deleterData() const
+    virtual SharedDataBlockInfo* sharedDataBlockInfo() const override
     {
-        return m_deleterData;
+        ASSERT_NOT_REACHED();
+        return nullptr;
     }
 
-    void reallocate(size_t newByteLength);
+    virtual void reallocate(size_t newByteLength) override;
 
     void* operator new(size_t size);
     void* operator new[](size_t size) = delete;
 
 private:
-    // Indicates whether the backing store was created for SharedArrayBuffer
-    bool m_isShared;
-    bool m_isAllocatedByPlatformAllocator;
+    NonSharedBackingStore(void* data, size_t byteLength, BackingStoreDeleterCallback callback, void* callbackData, bool isAllocatedByPlatform);
+
     void* m_data;
     size_t m_byteLength;
     BackingStoreDeleterCallback m_deleter;
     void* m_deleterData;
+    bool m_isAllocatedByPlatform;
+};
+
+#if defined(ENABLE_THREADING)
+class SharedDataBlockInfo {
+public:
+    SharedDataBlockInfo(void* data, size_t byteLength)
+        : m_data(data)
+        , m_byteLength(byteLength)
+        , m_refCount(0)
```

Comment:
`SharedArrayBufferObjectBackingStoreData's m_refCount` was initialized as 1, but `SharedDataBlockInfo's m_refCount` is initialized as 0. However, they have the same `deref` algorithm. is this right?

Reply:
Yes, I changed it to increase the counter when `SharedDataBlockInfo` is allocated in each `SharedBackingStore`.
Because `SharedDataBlockInfo` is valid only when it is shared by `SharedBackingStore`.
In `deref` algorithm, I check everytime that it has a valid counter (> 0) to ensure that this `deref` call is safe.